crates/compiler/src/listeners/compiler_listener.rs

```diff
@@ -0,0 +1,200 @@
+//! Adapted from the listener part of <https://github.com/YarnSpinnerTool/YarnSpinner/blob/da39c7195107d8211f21c263e4084f773b84eaff/YarnSpinner.Compiler/Compiler.cs>
+
+use crate::prelude::*;
+use antlr_rust::parser_rule_context::ParserRuleContext;
+use antlr_rust::token::Token;
+use antlr_rust::tree::{ParseTreeListener, ParseTreeVisitorCompat};
+use rusty_yarn_spinner_core::prelude::*;
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::rc::Rc;
+
+mod emit;
+use crate::parser::generated::yarnspinnerparser::{
+    BodyContext, HeaderContext, NodeContext, YarnSpinnerParserContextType,
+};
+use crate::prelude::generated::yarnspinnerparser::BodyContextAttrs;
+use crate::prelude::generated::yarnspinnerparserlistener::YarnSpinnerParserListener;
+use crate::visitors::CodeGenerationVisitor;
+pub(crate) use emit::*;
+use rusty_yarn_spinner_core::prelude::instruction::OpCode;
+
+pub(crate) struct CompilerListener<'input> {
+    pub(crate) debug_infos: Rc<RefCell<Vec<DebugInfo>>>,
+    /// The program being generated by the compiler.
+    pub(crate) program: Rc<RefCell<Program>>,
+    /// the list of nodes we have to ensure we track visitation
+    pub(crate) tracking_nodes: Rc<RefCell<HashSet<String>>>,
+    pub(crate) diagnostics: Rc<RefCell<Vec<Diagnostic>>>,
+    /// The current node to which instructions are being added.
+    current_node: Option<Node>,
+    /// The current debug information that describes [`current_node`].
+    current_debug_info: DebugInfo,
+    /// Whether we are currently parsing the
+    /// current node as a 'raw text' node, or as a fully syntactic node.
+    raw_text_node: bool,
+    file: FileParseResult<'input>,
+    label_count: usize,
+}
+
+impl<'input> CompilerListener<'input> {
+    pub(crate) fn new(tracking_nodes: HashSet<String>, file: FileParseResult<'input>) -> Self {
+        Self {
+            file,
+            tracking_nodes: Rc::new(RefCell::new(tracking_nodes)),
+            current_node: Default::default(),
+            current_debug_info: Default::default(),
+            raw_text_node: Default::default(),
+            diagnostics: Default::default(),
+            program: Default::default(),
+            label_count: Default::default(),
+            debug_infos: Default::default(),
+        }
+    }
+
+    /// Generates a unique label name to use in the program.
+    ///
+    /// ## Params
+    /// - `commentary` Any additional text to append to the
```

Comment:
move the next line onto this one

Reply:
[`cef2d10` (#56)](https://github.com/Mafii/rusty-yarn-spinner/pull/56/commits/cef2d10c94e516b586cb08143d5f31d26aebe7b5)

Reply:
[`cef2d10` (#56)](https://github.com/Mafii/rusty-yarn-spinner/pull/56/commits/cef2d10c94e516b586cb08143d5f31d26aebe7b5)